src/Aspire.Hosting/Orchestrator/ApplicationOrchestrator.cs

```diff
@@ -279,54 +362,47 @@ public async Task StopResourceAsync(string resourceName, CancellationToken cance
                 State = state,
                 StartTimeStamp = startTimeStamp,
-                StopTimeStamp = stopTimeStamp,
-                Properties = s.Properties.SetResourceProperty(KnownProperties.Resource.ParentName, parentName)
+                StopTimeStamp = stopTimeStamp
             }).ConfigureAwait(false);
 
-            // the parent name needs to be an instance name, not the resource name.
-            // parent the children of the child under the first resource instance.
-            await SetChildResourceAsync(child, child.GetResolvedResourceNames()[0], state, startTimeStamp, stopTimeStamp)
+            // Recurse to set the child resources of the child.
+            await SetChildResourceAsync(child, state, startTimeStamp, stopTimeStamp)
                 .ConfigureAwait(false);
         }
     }
 
-    private async Task SetExecutableChildResourceAsync(IResource resource)
-    {
-        // the parent name needs to be an instance name, not the resource name.
-        // parent the children under the first resource instance.
-        var parentName = resource.GetResolvedResourceNames()[0];
-
-        foreach (var child in _parentChildLookup[resource])
-        {
-            await _notificationService.PublishUpdateAsync(child, s => s with
-            {
-                Properties = s.Properties.SetResourceProperty(KnownProperties.Resource.ParentName, parentName)
-            }).ConfigureAwait(false);
-
-            await SetExecutableChildResourceAsync(child).ConfigureAwait(false);
-        }
-    }
-
-    private async Task PublishResourcesWithInitialStateAsync()
+    private async Task PublishResourcesInitialStateAsync(CancellationToken cancellationToken)
     {
         // Publish the initial state of the resources that have a snapshot annotation.
         foreach (var resource in _model.Resources)
         {
+            // Process relationships for the resource.
+            var relationships = ApplicationModel.ResourceSnapshotBuilder.BuildRelationships(resource);
+            var parent = resource is IResourceWithParent hasParent
+                ? hasParent.Parent
+                : resource.Annotations.OfType<ResourceRelationshipAnnotation>().LastOrDefault(r => r.Type == KnownRelationshipTypes.Parent)?.Resource;
+
             await _notificationService.PublishUpdateAsync(resource, s =>
             {
                 return s with
                 {
+                    Relationships = relationships,
+                    Properties = parent is null ? s.Properties : s.Properties.SetResourceProperty(KnownProperties.Resource.ParentName, parent.GetResolvedResourceNames()[0]),
                     HealthReports = GetInitialHealthReports(resource)
                 };
             }).ConfigureAwait(false);
+
+            // Notify resources that they need to initialize themselves.
+            var initializeEvent = new InitializeResourceEvent(resource, _eventing, _loggerService, _notificationService, _serviceProvider);
```

Comment:
Could this be limiting? What if we wanted to support a scenario where we add resources to the app model after we loop through this list here. We might be getting to a point where we need to consider that. if we are going to start introducing a place to bootstrap resource lifecycle then maybe now is the time.

Reply:
Upon reflection I think this is fine.

Reply:
@davidfowl and I spoke about this and posited that it's safe to manipulate the model up until the before start event. After that point, model changes result in undefined behavior.